File: gosdk-develop/lib/rpcRequest.go
package lib

import (
	"fmt"
	"github.com/milon-labs/milon-go-sdk/postcard"
)

const ContentTypeMilonPostcard = "application/x-milon+postcard"
const ContentTypeMilonJson = "application/x-milon+json"

type MethodType uint16

func (mt *MethodType) MarshalPostcard(serializer *postcard.Serializer) error {
	return serializer.SerializeU16(uint16(*mt))
}

func (mt *MethodType) UnmarshalPostcard(deserializer *postcard.Deserializer) error {
	value, err := deserializer.DeserializeU16()
	if err != nil {
		return fmt.Errorf("failed to deserialize MethodType: %w", err)
	}
	*mt = MethodType(value)
	return nil
}

const (
	// --- Core chain / execute (1-49, step 5)
	MethodTypeChainHead      MethodType = 1
	MethodTypeSubmitTx       MethodType = 5
	MethodTypeSimulateTx     MethodType = 10
	MethodTypeView           MethodType = 15
	MethodTypeGetAccount     MethodType = 20
	MethodTypeEventsByTxHash MethodType = 25

	// --- Query block / tx (50-149, step 5) -----------------------------
	MethodTypeGetBlockByHeight  MethodType = 50
	MethodTypeGetTxByHash       MethodType = 55
	MethodTypeGetTxHistoryProof MethodType = 60

	// --- Resource raw / batch (150-199, step 5) ------------------------
	MethodTypeGetResource                MethodType = 150
	MethodTypeGetResourcePathByHash      MethodType = 155
	MethodTypeBatchGetResourcePathByHash MethodType = 160
	MethodTypeGetAccessValue             MethodType = 165
)

func NewRpcRequest(method MethodType, requestId RequestID, body []byte) *RpcRequest {
	tx := &RpcRequest{
		Method:    method,
		RequestId: requestId,
		Body:      body,
	}

	return tx
}

type RpcRequest struct {
	Method    MethodType
	RequestId RequestID
	Body      []byte
}

func (req *RpcRequest) MarshalPostcard(serializer *postcard.Serializer) error {
	var err error

	if err = req.Method.MarshalPostcard(serializer); err != nil {
		return fmt.Errorf("failed to serialize Method: %w", err)
	}

	if err = serializer.SerializeU64(uint64(req.RequestId)); err != nil {
		return fmt.Errorf("failed to serialize RequestId: %w", err)
	}

	if err = serializer.SerializeBytes(req.Body); err != nil {
		return fmt.Errorf("failed to serialize Body: %w", err)
	}

	return nil
}

func (req *RpcRequest) UnmarshalPostcard(deserializer *postcard.Deserializer) error {
	var err error

	if err = req.Method.UnmarshalPostcard(deserializer); err != nil {
		return fmt.Errorf("failed to deserialize Method: %w", err)
	}

	requestId, err := deserializer.DeserializeU64()
	if err != nil {
		return fmt.Errorf("failed to deserialize RequestId: %w", err)
	}
	req.RequestId = RequestID(requestId)

	req.Body, err = deserializer.DeserializeBytes()
	if err != nil {
		return fmt.Errorf("failed to deserialize Body: %w", err)
	}
	if req.Body == nil {
		req.Body = []byte{}
	}

	return nil
}

File: gosdk-develop/python/milon_sdk/lib/rpc_request.py
"""RPC 请求线格式（对应 Go lib/rpcRequest.go）。

Postcard 线（Content-Type: application/x-milon+postcard）：
    method(u16 变长) + request_id(u64 变长) + body(bytes)

MethodType 编号（1-165，与 Go 常量一一对应）：
    1=ChainHead 5=SubmitTx 10=SimulateTx 15=View 20=GetAccount 25=EventsByTxHash
    50=GetBlockByHeight 55=GetTxByHash 60=GetTxHistoryProof
    150=GetResource 155=GetResourcePathByHash 160=BatchGetResourcePathByHash 165=GetAccessValue
"""
from __future__ import annotations

from enum import IntEnum

from ..postcard import Deserializer, Serializer

CONTENT_TYPE_MILON_POSTCARD = "application/x-milon+postcard"
CONTENT_TYPE_MILON_JSON = "application/x-milon+json"


class MethodType(IntEnum):
    CHAIN_HEAD = 1
    SUBMIT_TX = 5
    SIMULATE_TX = 10
    VIEW = 15
    GET_ACCOUNT = 20
    EVENTS_BY_TX_HASH = 25
    GET_BLOCK_BY_HEIGHT = 50
    GET_TX_BY_HASH = 55
    GET_TX_HISTORY_PROOF = 60
    GET_RESOURCE = 150
    GET_RESOURCE_PATH_BY_HASH = 155
    BATCH_GET_RESOURCE_PATH_BY_HASH = 160
    GET_ACCESS_VALUE = 165


class RpcRequest:
    __slots__ = ("method", "request_id", "body")

    def __init__(self, method: MethodType, request_id: int, body: bytes):
        self.method = MethodType(method)
        self.request_id = request_id
        self.body = bytes(body)

    def marshal_postcard(self, serializer: Serializer) -> None:
        serializer.serialize_u16(int(self.method))
        serializer.serialize_u64(self.request_id)
        serializer.serialize_bytes(self.body)

    @classmethod
    def unmarshal_postcard(cls, d: Deserializer) -> "RpcRequest":
        method = MethodType(d.deserialize_u16())
        request_id = d.deserialize_u64()
        body = d.deserialize_bytes()
        return cls(method, request_id, body)
